### PR TITLE
Refactor resource fitness functions and Lexicographic mode of GA

### PR DESCRIPTION
**Pull request by Timotshak**
- Add resource types specification in resource fitness functions
 - Introduce new resource fitness function which summarizes resources usage
 - Change logic of Lexicographic mode of GA to set
 fitness function for optimizization after reaching deadline

### CODE REVIEW

File: sampo/scheduler/genetic/schedule_builder.py [hunk 1]
```diff
@@ -189,7 +186,7 @@ def build_schedule(wg: WorkGraph,
         # save best individuals
         hof = tools.HallOfFame(1, similar=compare_individuals)
 
-        fitness_f = fitness_constructor(native.evaluate)
+        fitness_f = fitness_constructor(native.evaluate) if not have_deadline else TimeFitness(native.evaluate)
```
Comment (StannisMod) [inline sampo/scheduler/genetic/schedule_builder.py:189]: `fitness_constructor` is already parametrization about fitness function.
You shouldn't have any overrides of it. 
If you are going to do something like this, do it in `GeneticScheduler` class.

File: sampo/utilities/resource_usage.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    usage = defaultdict(lambda: np.zeros_like(points))
+
+    is_none = resources_names is None
+    resources_names = set(resources_names) if not is_none else {}
+
+    for swork in schedule.works:
+        start = points.bisect_left(swork.start_time)
+        finish = points.bisect_left(swork.finish_time + 1)
+        for worker in swork.workers:
+            if worker.name in resources_names or is_none:
```
Comment (StannisMod) [inline sampo/utilities/resource_usage.py:24]: If `resource_names` is `None`, you receives `TypeError` here

**Review by StannisMod — CHANGES_REQUESTED** [review]
Changes requested

File: sampo/scheduler/genetic/schedule_builder.py (hunk 1 above)
Comment (Timotshak) [reply sampo/scheduler/genetic/schedule_builder.py:189]: if a deadline is given, then a two-stage lexicographic genetic algorithm is applied. Therefore, at the first stage, optimization is carried out over Time, and already at the second stage, the function passed to fitness_constructor is optimized.

File: sampo/utilities/resource_usage.py (hunk 2 above)
Comment (Timotshak) [reply sampo/utilities/resource_usage.py:24]: no, because in this case resource_names will be initialized to an empty set

**Review by StannisMod — APPROVED** [review]
Approved

**Pull request merged**